File: 15-Flyweight/src/go/tree_type.go
package main

import "fmt"

// TreeType is the Flyweight — stores ONLY intrinsic (shared) state.
// Fields are unexported and never mutated, so it is safe to share.
// Extrinsic state (x, y) is passed IN via Draw().
type TreeType struct {
	name    string
	color   string
	texture string
}

func (t *TreeType) Draw(x, y int) {
	fmt.Printf("  Drawing %s (%s, texture=%s) at (%d, %d)\n",
		t.name, t.color, t.texture, x, y)
}

func (t *TreeType) Name() string { return t.name }

File: 15-Flyweight/src/rust/tree_type.rs
/// Flyweight — stores ONLY intrinsic (shared) state.
/// Immutable by default in Rust; shared via `Rc` (see the factory).
/// Extrinsic state (x, y) is passed IN via `draw`.
pub struct TreeType {
    name: String,
    color: String,
    texture: String,
}

impl TreeType {
    pub fn new(name: String, color: String, texture: String) -> Self {
        TreeType { name, color, texture }
    }

    pub fn draw(&self, x: i32, y: i32) {
        println!(
            "  Drawing {} ({}, texture={}) at ({}, {})",
            self.name, self.color, self.texture, x, y
        );
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}
